File: tests/fixtures/kotlin/user_service.kt
package com.example.demo.service

import java.io.IOException

/**
 * Primary user-facing service with CRUD + auth.
 *
 * Used by the example suite as a canonical smoke fixture for the
 * Kotlin adapter — exercises visibility modifiers, KDoc, annotations,
 * generics, and nested types all at once.
 */
@Service
@Deprecated("use V2")
open class UserService(
    @JvmField val name: String,
    private val items: MutableList<String> = mutableListOf(),
) : BaseService(), UserRepository, AutoCloseable {

    companion object {
        const val MAX_USERS: Int = 100
        @JvmStatic
        fun defaultName(): String = "anonymous"
    }

    // package-private equivalent in Kotlin is `internal` — module-scoped
    internal var packagePrivateField: Int = 0

    protected val cache: MutableMap<String, Any> = mutableMapOf()

    /**
     * Persists the incoming user; throws IOException on I/O failure.
     */
    @Throws(IOException::class, IllegalArgumentException::class)
    @Override
    override fun save(user: String): Boolean {
        items.add(user)
        return true
    }

    private fun <T : Comparable<T>> findMax(xs: List<T>): T = xs.max()

    abstract fun compute(x: Int, y: Int): Int

    override fun close() {
        items.clear()
    }

    /** Nested class — public by default in Kotlin. */
    class Inner {
        fun value(): Int = 42
        constructor() { }
    }

    /** Nested interface (fun interface). */
    fun interface Callback {
        fun onDone(code: Int): Unit
    }
}

File: tests/fixtures/scala/user_service.scala
package com.example.demo.service

import java.io.IOException
import scala.collection.mutable

/**
 * Primary user-facing service with CRUD + auth.
 *
 * Canonical smoke fixture for the Scala adapter: exercises access
 * modifiers, Scaladoc, annotations, generics, nested types, and both
 * val- and var-defined primary-ctor properties.
 */
@deprecated("use V2")
@SerialVersionUID(1L)
class UserService(
  val name: String,
  private val items: mutable.ListBuffer[String] = mutable.ListBuffer.empty,
) extends BaseService with UserRepository with AutoCloseable {

  final val MAX_USERS: Int = 100

  // Scala's `internal`-like default is `public` — no modifier is public.
  var packageDefault: Int = 0

  protected val cache: mutable.Map[String, Any] = mutable.Map.empty

  /**
   * Persists the incoming user; throws IOException on I/O failure.
   */
  @throws(classOf[IOException])
  override def save(user: String): Boolean = {
    items += user
    true
  }

  private def findMax[T: Ordering](xs: List[T]): T = xs.max

  def compute(x: Int, y: Int): Int

  override def close(): Unit = items.clear()

  /** Nested class — public by default. */
  class Inner {
    def value: Int = 42
  }

  /** Nested trait. */
  trait Callback {
    def onDone(code: Int): Unit
  }
}

/** Companion object — standard Scala singleton-paired-with-class idiom. */
object UserService {
  def apply(name: String): UserService = new UserService(name)
}
